uint128.go

```diff
@@ -100,41 +100,59 @@ func (u Uint128) Xor64(v uint64) Uint128 {
 
 // Add returns u+v.
 func (u Uint128) Add(v Uint128) Uint128 {
+	if u.Cmp(Max.Sub(v)) == 1 {
+		panic("Value being added is greater than max allowable number, causes overflow")
+	}
 	lo, carry := bits.Add64(u.Lo, v.Lo, 0)
 	hi, _ := bits.Add64(u.Hi, v.Hi, carry)
 	return Uint128{lo, hi}
 }
 
 // Add64 returns u+v.
 func (u Uint128) Add64(v uint64) Uint128 {
+	if u.Cmp(Max.Sub64(v)) == 1 {
+		panic("Value being added is greater than max allowable number, causes overflow")
+	}
 	lo, carry := bits.Add64(u.Lo, v, 0)
 	hi := u.Hi + carry
 	return Uint128{lo, hi}
 }
 
 // Sub returns u-v.
 func (u Uint128) Sub(v Uint128) Uint128 {
+	if u.Cmp(v) == -1 {
+		panic("Value being subtracted is greater than the number, causes underflow")
+	}
```

Comment:
Here it could be better because if the cmp result is -1, we don't proceed further and from what I gather, Cmp is O(1) and so are the subsequent subtractions.

Reply:
Well, if we assume that the common case is *not* underflowing, then it's more efficient to inspect the `borrow` returned from `bits.Sub64` (which has to be computed anyway) rather than doing a separate set of comparisons via `u.Cmp`.

Honestly though the performance difference is probably moot; I'd just prefer to check the `borrow` for aesthetic reasons (symmetry with `Add`).